### ISSUE TITLE
Option to pass in dom element

### ISSUE THREAD

**Issue by SyedWasiHaider** [tscanlin/tocbot#99]
I'm using marked to generate HTML from markdown. The problem is that the html has not rendered or is not present in the dom when toc.init is called.

Currently my workaround is to use a timeout but its hacky
    setTimeout((() => toc.refresh()), 2000)

Would it be possible to pass in a dom element to toc so that instead of querying document, it queries the dom element which will have the headers with ids.

**Comment by tscanlin** [comment]
That’s what contentSelector does.

Do you get a callback from the function that generates the html from markdown? That’s what you’d need. Or just render serverside

**Comment by SyedWasiHaider** [comment]
Sorry let me be a bit more clear:

I see here: https://github.com/tscanlin/tocbot/blob/a06e4df3c36fc5e6eb7722ad640b878049a6be51/src/js/parse-content.js#L97

that document.querySelector is used to find the headers. If I could instead pass in my own dom element to call querySelector on that would be great.

If it helps I'm essentially doing something like this:

```
readme = require('readme.md')
div = document.createElement('div')
div.innerHTML = readme.trim()
# Here tocbot could query this div for the headers
```

I'm using https://github.com/peerigon/markdown-loader for loading that markdown file.

**Comment by SyedWasiHaider** [comment]
My understanding was that tocbot essentially gets the headers from the document then constructs the  nice table of contents element. So technically the headers don't need to be in the document. Feel free to close this issue if I'm not making sense.

**Comment by tscanlin** [comment]
So you want to read the headers from the element before it’s added to the Dom? Does requiring the readme.md like that actually work?

Adding the content async like that is not recommended because when users come back to the page with a hash link they won’t go to the right heading since the content isn’t ready yet.

**Comment by SyedWasiHaider** [comment]
Requiring the readme.md does work (using the plugin I mentioned above and the renderer from Markedjs).

I think the problem is I'm trying to do it all on the client side (because I'm generating some docs that are specific to the user). With regards to the hash link issue, yeah I would have to wait for the content to render.

**Comment by tscanlin** [comment]
Why not just add the content in a hidden div with display none? Or what’s the purpose of not attaching the Dom node to the document?

**Comment by SyedWasiHaider** [comment]
Unfortunately I'm using this crappy framework called AngularJS (not my choice). There is no straightforward way to determine a dom element has rendered. I could use some timeout hack or polling mechanism but I really don't want to.

**Comment by tscanlin** [comment]
You just need to append the div to body before you call tocbot.init() and you should be good.

**Comment by SyedWasiHaider** [comment]
ok indeed at least the sidebar to show up now but appending to the dom with display:none. I was just trying to avoid having to do that. Only issue now is that the url is relative to the root rather than the current window location...

**Comment by tscanlin** [comment]
What do you mean?

**Comment by SyedWasiHaider** [comment]
Sorry I updated my comment.

**Comment by tscanlin** [comment]
Can you link me to an example or include a screenshot?

**Comment by SyedWasiHaider** [comment]
Unfortunately I cannot. I think your library is doing what is supposed to and changing the hash of the url as expected. I just need to update my routing logic to not refresh on a hash change.

**Comment by dkniffin** [comment]
@tscanlin Sorry to ping on an old issue, but I'd like to request this feature as well, for different reasons. 

I'm using [StimulusJS](https://stimulus.hotwire.dev/). The way it works is I'd create a controller class, and it handles the linking between the html and the JS. Because of that, passing a selector to tocbot is both overkill and it would likely cause bugs because it could be initialized multiple times.

If you'd like, I can create a PR for this.

**Comment by tscanlin** [comment]
Lol, I finally got around to reviewing / merging the PR for this. Thanks @dkniffin!
Closing this now.

**Comment by tscanlin** [comment]
published in ` tocbot@4.18.0`
